src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -561,10 +563,42 @@ void SubqueryToJoinRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node)
     return;
   }
 
-  const auto join_node = JoinNode::make(predicate_node_info->join_mode, join_predicates);
+  const auto join_mode = predicate_node_info->join_mode;
+  const auto join_node = JoinNode::make(join_mode, join_predicates);
   lqp_replace_node(node, join_node);
   join_node->set_right_input(pull_up_result.adapted_lqp);
 
+  if (pull_up_result.adapted_lqp) {
+    const auto& estimator = cost_estimator->cardinality_estimator;
+    const auto probe_side_cardinality = estimator->estimate_cardinality(join_node->left_input());
+    const auto build_side_cardinality = estimator->estimate_cardinality(join_node->right_input());
+    const auto& primary_join_predicate = join_predicates[0];
+
+    bool primary_join_predicate_is_equals = false;
+    if (const auto predicate_expression =
+            std::dynamic_pointer_cast<BinaryPredicateExpression>(primary_join_predicate)) {
+      if (predicate_expression->predicate_condition == PredicateCondition::Equals) {
+        primary_join_predicate_is_equals = true;
+      }
+    }
+
+    if ((join_mode == JoinMode::Semi || join_mode == JoinMode::AntiNullAsTrue ||
+         join_mode == JoinMode::AntiNullAsFalse) &&
+        build_side_cardinality > probe_side_cardinality * 10 && build_side_cardinality > 1'000 &&
+        primary_join_predicate_is_equals) {
+      // Semi/Anti joins are currently handled by the hash join, which performs badly if the right side is much
+      // bigger than the left side. For that case, we add a second semi join on the right side, which throws out all
+      // values that will not be found by the primary (first) predicate of the later join, anyway. However, we can only
+      // throw away values on the build side if the primary predicate is an equals predicate. For an example, see TPC-H
+      // query 21. That query is the main reason this part exists. The current thresholds are somewhat arbitrary. If we
+      // start to see more of those cases, we can try and find a more complex heuristic. We call this "semi join
```

Comment:
To make it clearer for anybody reading that in two years, maybe write something like "We add an additional join operation to the plan that behaves similar to a semi join reduction."?

Reply:
I find "For that case, we add a second semi join on the build side, which throws out all values that will not be found by the primary (first) predicate of the later join, anyway." clearer that "semi join reduction", which I have not heard of before you mentioned the term. Added a link to a paper that explains it.